src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -319,37 +779,274 @@ pub fn sort_hack_2(scene:&Scene) {
+///
+/// Example: We need to split on the `SideLine2` and highlight the shapes closer to the
+/// output port. That means we need to do the geometric split on  `Corner2`, `SideLine2`, `Corner3`,
+/// which we can access via `split_shapes` and apply the highlighting to `SideLine` `Corner` and
+/// `MainLine`/`Arrow`, which we can access via `output_side_shapes`. The remaining shapes that must
+/// not be highlighted can be accessed via `input_side_shapes`.
+#[derive(Clone,Debug)]
+struct SemanticSplit {
+    /// Ids of the shapes in the order they appear in the edge. Shapes that fill the same "slot" in
+    /// the shape and must be handled together, are binned into a sub-vector. That can be the case
+    /// for shapes that are present in the back and the front of the shape.
+    ordered_part_ids : Vec<Vec<display::object::Id>>,
+    /// The index the shape where the edge split occurs in the `ordered_part_ids`.
+    split_index      : usize,
+}
+
+impl SemanticSplit {
+
+    fn new(edge_data:&EdgeModelData, split_shape:display::object::Id) -> Option<Self> {
+        let ordered_part_ids = Self::semantically_binned_edges(edge_data);
+
+        // Find the object id in our `ordered_part_ids`
+        let mut split_index  = None;
+        for (index, shape_ids) in ordered_part_ids.iter().enumerate() {
+            if shape_ids.contains(&split_shape) {
+                split_index = Some(index);
+                break
+            }
+        }
+        let split_index = split_index?;
+
+        Some(SemanticSplit {ordered_part_ids,split_index})
+    }
+
+    /// Return a ordered vector that contains the ids of the shapes in the order they appear in the
+    /// edge. Shapes that are to be handled as in the same place, are binned into a sub-vector.
+    /// This enables us to infer which parts are next to each other, and which ones are
+    /// "source-side"/"target-side".
+    ///
+    /// In general, we treat the equivalent shape from front and back as the same, as they tend to
+    /// occupy the same space within the shape and thus need to be handled together. But,
+    /// for example, also the arrow needs to be handled together with the main line.
+    ///
+    /// The important information we create here is the rough adjacency of shapes. This is used to
+    /// determine which shapes are adjacent to avoid rendering a split on a shape that can be all
+    /// focus on or all focus off.
+    fn semantically_binned_edges(edge_data:&EdgeModelData) -> Vec<Vec<display::object::Id>> {
+        let front  = &edge_data.front;
+        let back   = &edge_data.back;
+        let layout = edge_data.layout_state.get();
+        match layout {
+            LayoutState::DownLeft | LayoutState::DownRight => {
+                vec![
+                    vec![EdgeShape::id(&front.side_line),  EdgeShape::id(&back.side_line)  ],
+                    vec![EdgeShape::id(&front.corner),     EdgeShape::id(&back.corner)     ],
```

Comment:
I was writing in my previous PR that this is incorrect. `front.corner` and `back.corner` do not "occupy the same space", as docs tell. Front corner is the part that is visible next to node, while back corner is a very short part which fills the hole between the shape and `front.corner`. They are visible at the same time, always. - see the definition of the shapes - they differ . One is `ring - node_shape` and one is `ring * node_shape`.

Reply:
I've added a note about this. It doesn't really matter as there is no way a focus split can create a self-intersection on those shapes. We can just handle them as part of the same  shape group.